pkg/services/ngalert/store/alert_rule.go

```diff
@@ -553,23 +549,28 @@ func (st DBstore) GetAlertRulesForScheduling(ctx context.Context, query *ngmodel
 				}
 			}
 			rules = append(rules, rule)
+
+			// TODO append only if folder is not present
+			foldersPerOrg[rule.OrgID] = append(foldersPerOrg[rule.OrgID], rule.NamespaceUID)
 		}
 
 		query.ResultRules = rules
 
 		if query.PopulateFolders {
-			foldersSql := sess.Table("folder").Alias("d").Select("d.uid, d.title, d.parent_uid").
-				Where(`EXISTS (SELECT 1 FROM alert_rule a WHERE d.uid = a.namespace_uid AND d.org_id = a.org_id)`)
-			if len(disabledOrgs) > 0 {
-				foldersSql.NotIn("org_id", disabledOrgs)
-			}
+			for org := range foldersPerOrg {
+				folders, err := st.FolderService.GetFolders(ctx, folder.GetFoldersQuery{
+					OrgID:        org,
+					UIDs:         foldersPerOrg[org],
+					WithFullpath: true,
+				})
+				if err != nil {
+					return fmt.Errorf("failed to fetch a list of folders that contain alert rules: %w", err)
+				}
 
-			if err := foldersSql.Find(&folders); err != nil {
-				return fmt.Errorf("failed to fetch a list of folders that contain alert rules: %w", err)
-			}
-			query.ResultFoldersTitles = make(map[string]string, len(folders))
-			for _, folder := range folders {
-				query.ResultFoldersTitles[folder.Uid] = ngmodels.GetNamespaceKey(folder.ParentUid, folder.Title)
+				query.ResultFoldersTitles = make(map[string]string)
+				for _, folder := range folders {
+					query.ResultFoldersTitles[folder.UID] = folder.Fullpath
```

Comment:
we use to map folder UIDs to title (existing behaviour) but given that folders are uniquely identified by their `org_id` and `uid` we have to include the org in the key as well

Reply:
Although that seems to be an obvious solution in this case, I am afraid that this could be a bit error-prone in a large environment without proper batching of queries (see my other comment). 
An alternative could be a parameter for `GetFoldersQuery`, let's say `Filter`, that is an interface with one single method `SQLFilter(folderAlias string) string`. So it would let us inject a filter by, for example, all folders that have alert rules.

Reply:
What would be the benefit of the `Filter`? I don't fully understand. I would hesitate to give a lot of freedom to the callers otherwise the final query can vary a lot and it will be hard to compare its performance and spot issues (as it happens with the search query today that it has similar filters).